### PR TITLE
Major setup proposal

### PR DESCRIPTION
**Pull request by ofloveandhate**
Initial file structure, Float class implementation, documentation, build configuration using autotools, autotesting using Boost.UnitTest, eigen dependency, and other setup pieces.  I propose this pull request, so that other developers can play in the ecosystem we are creating, even though this commit has an 85% functional Float class.

Note deficiencies in the Float class are the lack of MPI send/receive, and the lack of serialization support.  Next work will resolve these current shortcomings.

### CODE REVIEW

**Comment by jbcolli2** [comment]
Brake,
   Ok, I'm feeling dumb here but I'm having trouble with autoreconf.  I got the autotools from brew, but when I run autoreconf in the core directory is gives me errors, saying I may be missing certain macros.  Is there a place to get a more updated version of autoreconf?  Thanks

**Comment by matthewniemerg** [comment]
Jeb,

The version of autoconf that I have is 2.69 and everything works for me when I clone this pull and in the core directory run `autoreconf - i'.  You can download the source for autoconf from http://ftp.gnu.org/gnu/autoconf/ -- just choose latest version, down towards the bottom.

Matt

**Comment by timothodge** [comment]
Jeb,
Depending on what Macros it says you are missing you may need to install
pkgconfig. I did this using brew and then autoreconf -i worked.

Best,
Tim

On Thu, Feb 26, 2015 at 3:00 PM, matthewniemerg <notifications@github.com>
wrote:

> Jeb,
>
> The version of autoconf that I have is 2.69 and everything works for me
> when I clone this pull and in the core directory run `autoreconf - i'. You
> can download the source for autoconf from http://ftp.gnu.org/gnu/autoconf/
> -- just choose latest version, down towards the bottom.
>
> Matt
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/danbates/b2/pull/5#issuecomment-76283577>.
>



-- 
Tim Hodges
Colorado State University
Department of Mathematics

**Comment by timothodge** [comment]
Dear all,

I was able to autoreconf -i in the core directory. When I try to make after
autoreconf -i I get* ./include/float.hpp:2338:3: error: explicit
specialization of non-template 'Eigen::NumTraits'* at the bottom of all
that is thrown to the screen when I attempt to 'make'.

At the top of this attempt I get

* ./include/float.hpp:21568:38: error: 'isnan' was not declared in this
scope*
Hope I am the only one with this error and any help would be appreciated.

Best,
Tim

On Thu, Feb 26, 2015 at 4:44 PM, Tim Hodges <hodges@math.colostate.edu>
wrote:

> Jeb,
> Depending on what Macros it says you are missing you may need to install
> pkgconfig. I did this using brew and then autoreconf -i worked.
>
> Best,
> Tim
>
> On Thu, Feb 26, 2015 at 3:00 PM, matthewniemerg <notifications@github.com>
> wrote:
>
>> Jeb,
>>
>> The version of autoconf that I have is 2.69 and everything works for me
>> when I clone this pull and in the core directory run `autoreconf - i'. You
>> can download the source for autoconf from
>> http://ftp.gnu.org/gnu/autoconf/ -- just choose latest version, down
>> towards the bottom.
>>
>> Matt
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/danbates/b2/pull/5#issuecomment-76283577>.
>>
>
>
>
> --
> Tim Hodges
> Colorado State University
> Department of Mathematics
>



-- 
Tim Hodges
Colorado State University
Department of Mathematics

**Comment by matthewniemerg** [comment]
Hi Tim,

I encountered this error earlier today.  It's a 'scope' issue with the isnan function (and similarly with isinf).  The fix is rather easy:  on those lines, just add 'std::' in front, and the scoping issues go away!

Matt

**Comment by timothodge** [comment]
Hey Matt,

Thank you for the help with the scope! I am still getting an error that
says *./include/float.hpp:2337:21: error: ‘NumTraits’ is not a class
template. *Any ideas on how to get rid of this?

Best,
Tim

On Thu, Feb 26, 2015 at 6:57 PM, matthewniemerg <notifications@github.com>
wrote:

> Hi Tim,
>
> I encountered this error earlier today. It's a 'scope' issue with the
> isnan function (and similarly with isinf). The fix is rather easy: on those
> lines, just add 'std::' in front, and the scoping issues go away!
>
> Matt
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/danbates/b2/pull/5#issuecomment-76317051>.
>



-- 
Tim Hodges
Colorado State University
Department of Mathematics

**Comment by ofloveandhate** [comment]
tim, wanna skype later?  i searched, and can't find any useful info about
this ‘NumTraits’ is not a class
template error.  initially, i am curious what version of what compiler you
are using.

**Comment by ofloveandhate** [comment]
i have made a commit fixing the std:: scoping of isnan and isinf.

**Comment by ofloveandhate** [comment]
i committed, removing an unnecessary #include, which should bring the
minimum boost version requirement back to 1.54.  I also enforced the
presence of Eigen, and removed a bad macro from configure.ac.

Tim, in particular, I am curious if these things fix your issue?

**Comment by aliddell** [comment]
As of this latest commit, make fails for me on Ubuntu with
/usr/include/boost/serialization/access.hpp:118:9: error: ‘class boost::multiprecision::number<boost::multiprecision::backends::mpfr_float_backend<0u, (boost::multiprecision::mpfr_allocation_type)1u> >’ has no member named ‘serialize’ (I do have libboost-serialization-dev package installed); succeeds on Fedora.

**Comment by ofloveandhate** [comment]
i'll try building in ubuntu, and get back to you.

**Comment by aliddell** [comment]
No need. Build succeeds with boost 1.55 on Ubuntu.

**Comment by ofloveandhate** [comment]
perfect.

**Comment by ofloveandhate** [comment]
can confirm 1.54 boost fails on ubuntu, 1.55 succeeds.

can every one report in with what OS and compiler they have built and
tested this with?

**Comment by ofloveandhate** [comment]
Macintosh OSX 10.10, Clang.  Boost 1.57.

**Comment by timothodge** [comment]
That would be great Dan! I am free from 11-1 today(Colorado time). If today
doesn't work we can always Skype another time.

**Comment by aliddell** [comment]
Fedora 21: `g++ --version` ==> g++ (GCC) 4.9.2 20150212 (Red Hat 4.9.2-6)
Ubuntu 14.04 LTS: `g++ --version` ==> g++ (Ubuntu 4.8.2-19ubuntu1) 4.8.2
Boost 1.55 on both platforms, as distributed in the repos

**Comment by jbcolli2** [comment]
Mac OSX 10.9.5, Clang, Boost 1.57.0

**Comment by matthewniemerg** [comment]
Mac OSX 10.9.5, compiled with mpicc and mpicxx version mpi3 which was itself compiled with gcc 4.7.3, boost 1.57.0

**Comment by danbates** [comment]
Trying to compile - Mac OSX 10.9.5, boost 1.57.0, gcc 4.9.2_1, MPFR 3.1.2-p10 (all brewed).

autoconf -i succeeds (with two errors: error: possibly undefined macro: AC_ENABLE_SHARED, ditto for STATIC).

./configure bails after failing to find MPFR: 

checking for library containing mpfr_get_version... no
configure: error: unable to find mpfr
Dans-MacBook-Pro-2:core bates$ brew upgrade mpfr
Error: mpfr 3.1.2-p10 already installed

Thoughts?  

FYI: Other ./configure issues:

...
checking whether g++ supports C++11 features with -std=gnu++11... yes
./configure: line 3810: AC_ENABLE_SHARED: command not found
./configure: line 3813: AC_ENABLE_STATIC: command not found
...
checking whether ln -s works... yes
./configure: line 4039: LT_INIT: command not found
...

**Comment by matthewniemerg** [comment]
Dan Ba.,

It looks as if you haven't set your CXXFLAGS and LDFLAGS to look for mpfr properly .

I used the following configure:

./configure MPICC=/usr/local/mpich3-gcc4.7.3/bin/mpicc MPICXX=/usr/local/mpich3-gcc4.7.3/bin/mpic++ CPPFLAGS="-m64 -I/usr/local/boost_1_57_0/include -I/usr/local/mpfr-3.1.2/include -I/usr/local/gmp-6.0.0/include -I/usr/local/include/eigen3" LDFLAGS="-L/usr/local/boost_1_57_0/lib -L/usr/local/mpfr-3.1.2/lib -L/usr/local/gmp-6.0.0/lib" --with-boost=/usr/local/boost_1_57_0

which worked. 

However, I was looking at the build and libtool, etc. still ended up using my CLANG compiler instead!  I just noticed this now.  When I add the CC and CXX macros on my configure, I get a linking error with some boost serialization bits.

For example, 

Undefined symbols for architecture x86_64:  "boost::archive::text_iarchive_impl<boost::archive::text_iarchive>::load(std::basic_string<char, std::char_traits<char>, std::allocator<char> >&)

--- and more linking errors, of a similar nature.  For others who may encounter this, a preliminary search reveals this SO thread as useful:  http://stackoverflow.com/questions/24718996/linker-error-undefined-symbols-stdstringc-str-const-on-macos-with-libboo .

Basically, there is a difference with clang libraries used and gcc's.  I'll see if this helps.

Matt

**Comment by ofloveandhate** [comment]
dan bates-
you must install libtool.  brew install libtool.

**Comment by jbcolli2** [comment]
Question: has anyone gotten the make to work yet.  I ask because mine seems to work fine, but I'm having issues running the test file and I wanted to know if anyone has gotten the whole shebang to work.  Thanks

**Comment by danbates** [comment]
Still having MPFR issues though brewing libtool eliminated my other
non-fatal errors.  Still no sign of MPFR (even if using Matt's ./configure
line).  I'll keep trying different things and will report back upon success.

Jeb: Not me...yet.

Thanks,
Dan

**Comment by ofloveandhate** [comment]
hey y'all

thanks for engaging in this first pull request, and trying to build the b2
core at this time.  your input is absolutely critical to our collective
success, and we all really appreciate everything you do, from brainstorming
and flowcharting, to building the software, to testing, and giving feedback
at every stage.

I have four things:

firstly, now seems like a good time to remind people that software must be
compiled using the same compiler used to compile the libraries against
which it links.

secondly, similarly, and specifically, i have had mediocre success with
trying to use gcc/g++ on my mac, because by default g++ and gcc wrap around
Apple's clang compiler when you install the command line tools from XCode,
and when installed, g++ and clang can fight with each other unless you
manage your $PATH and a few other environment variables.
If you want to use gcc/g++ on a mac, and don't want to use clang, you are
more than welcome to do that.  just know that any libraries you use *must*
have also been compiled using the same compiler.  the burden is on you to
ensure this, not the compiler, not ./configure, not brew, not macports, not
fink, not xcode, not anything other than yourself.  if you have multiple
versions of the same library on your machine, when you ./configure, use the
CPPFLAGS, LDFLAGS, and possibly set environment variables such as
LD_LIBRARY_PATH and LD_RUN_PATH.   There are also ways to specify the
location of things like boost.  If we need to add ways to specify gmp and
mpfr, we can do that, but you should be able to pass custom locations for
these by adding them to LDFLAGS and CPPFLAGS.  You may also specify the
compiler you want to use by passing the variable CXX when running
./configure, and eventually the MPI compiler via MPICXX.  when in doubt,
please first consult the help built-in to the b2/core/ configure script, by
going ./configure --help.  It's there to help you.  If it doesn't help you,
then we need to add more things to the configure.ac file, so that when the
script is built by autoreconf, the help gets in there.
the same goes if you have multiple compilers on the same machine.  setting
flags at ./configure time is how to get things to play nicely, but you must
ensure that everything you are linking together was build similarly.

thirdly, i remind you that when the b2 software is eventually distributed,
it will almost always be in compiled executable form, or tarball form.  The
only people who have to do the whole autoreconf thing are the developers.
 so stay comforted that the user will only have to go ./configure make, but
does not have to have the whole autotools set (autoconf, automake, libtool,
and sometimes pkg-config).  Our mission is to make sure that the
configure.ac file is complete and correct enough that the ./configure
script works for 99.9% of users; and that the code it builds is all correct
as well.

lastly, if you are having errors or trouble, please send *the entire
console output* including the exact errors, if you want help.  don't be
afraid to save it to a file and attach it, use pastebin, or other things.
 don't be afraid of long messages. the more specific you are, including how
you got where you are, what exact commands you issued, etc, the faster we
can help, and the less time we have to spend asking 'what did you do?' and
'what was the output?'.  compiler and runtime errors are extremely
specific, and we simply cannot diagnose 'problems' or 'errors' without
exact phrasing.  information like compiler version and operating system is
also really helpful.

thanks again for everything, including the compiling and setup challenges
in this first major pull request.  i genuinely mean it.

dan brake

**Comment by aliddell** [comment]
In case anyone has been wondering about Cygwin, I was able to successfully configure-make-test on Cygwin, with g++ 4.9.2, boost 1.55, and Eigen 3.4.2. Just be sure everything is up to date. One caveat: I had to go in and add two lines to 'configure' (it will tell you where if you try to run it), since apparently Cygwin's bash will choke on
`if [condition]; do [statements]; else [no statements]; fi`
If this gives you trouble, just add `:` between `else` and `fi`

I know we're not actively targeting Cygwin, but I thought you all would be interested to know.

**Comment by matthewniemerg** [comment]
I now have everything working with custom built gcc-4.8.4 (compiled originally from clang), custom built gmp/mpfr libraries (where these were compiled within the gcc-4.8.4 source directory and moved to custom directories), and boost 1.57.0 (built with gcc-4.8.4) on Mac OSX 10.9.5.  

All tests passed correctly after commenting out "#define BOOST_TEST_DYN_LINK" in class_test.cpp file, as I only have static libraries built.

I have everything documented for anyone else who wants to build everything this way, without the use of third-party installers (homebrew, fink, or macports) for Mac.

**Comment by danbates** [comment]
Mac OSX 10.9.5, with mpich2/3.1.3_1.

FYI, a previous posting from Matt did the trick.  Previously, I didn't
think to change all the paths.  Now it works, with:

./configure MPICC=/usr/local/Cellar/mpich2/3.1.3_1/bin/mpicc
MPICXX=/usr/local/Cellar/mpich2/3.1.3_1/bin/mpic++ CPPFLAGS="-m64
-I/usr/local/Cellar/boost/1.57.0/include
-I/usr/local/Cellar/mpfr/3.1.2-p10/include
-I/usr/local/Cellar/gmp/6.0.0a/include
-I/usr/local/Cellar/eigen/3.2.4/include"
LDFLAGS="-L/usr/local/Cellar/boost/1.57.0/lib
-L/usr/local/Cellar/mpfr/3.1.2-p10/lib -L/usr/local/Cellar/gmp/6.0.0a/lib"
--with-boost=/usr/local/Cellar/boost/1.57.0


Notice I used boost for everything (note the /usr/local/Cellar in all
paths).


Now for testing....

Dan

**Comment by ofloveandhate** [comment]
hey b2

i made some commits to my develop branch addressing class, method, and
member commenting, test names, and addressed one bug i found (via test
driven design ftw).  please pull from develop on my fork if you wish to
view or use the changes.

all the best,
daniel

**Comment by ofloveandhate** [comment]
The Float class was beautiful but terrible.  I have removed the class from my fork.  Now, present are the barest skeleton of bertini2, and a complex class (or at least the files for it).  This pull request still represents the structure of the repo we want for our dev process.  i recommend the acceptance of this pull request.
